P083: Add MQ5 sensor plugin

pecord opened the pull request:
I got an MQ-5 gas sensor off amazon and wanted to add it to my ESPEasy sensor.  

During my research I found a PR for MQ2 - https://github.com/letscontrolit/ESPEasy/pull/96

I used that for my starting point, but the values were specific to MQ2.  I got on github and typed in MQ5 and found another repo for all the sensors that had similar code.  I ended up using this code/values to adapt it to MQ5.

https://github.com/mdsiraj1992/Gassensors/blob/master/arduino%20codes/mq5_sensor/mq5_sensor.ino

giig1967g committed:
bugfix for sending event in case of gpio commands

giig1967g committed:
bug fix for state = -1 in JSON

was reporting 65535

giig1967g committed:
bug fix for gpio events for P009 and P019

giig1967g committed:
fixed string error

giig1967g committed:
optimizing PLUGIN_MONITOR call

Timing stats average time is now 1.5ms compared to 2.5ms of the previous version.

giig1967g committed:
fixed esp32 compile failure

TD-er committed:
[GPS] Add TinyGPSPlus library 1.0.2

[Source](https://github.com/mikalhart/TinyGPSPlus)

TD-er committed:
[GPS] Add TinyGPSPlus library 1.0.2

[Source](https://github.com/mikalhart/TinyGPSPlus)

TD-er committed:
[GPS] Improved precision

Applied PR from @qbazd
[lat, lng force math operations #47](https://github.com/mikalhart/TinyGPSPlus/pull/47)

TD-er committed:
[GPS] Improved precision

Applied PR from @qbazd
[lat, lng force math operations #47](https://github.com/mikalhart/TinyGPSPlus/pull/47)

TD-er committed:
[GPS] support for fix quality and fix mode

Applied this PR:
[support for fix quality and fix mode](https://github.com/mikalhart/TinyGPSPlus/pull/38)

TD-er committed:
[GPS] support for fix quality and fix mode

Applied this PR:
[support for fix quality and fix mode](https://github.com/mikalhart/TinyGPSPlus/pull/38)

giig1967g committed:
optimized monitor call

Execution time reduced to1ms average

TD-er committed:
[GPS] New plugin P082 GPS - initial commit

TD-er committed:
[GPS] New plugin P082 GPS - initial commit

CaptainRoot committed:
Update _P076_HLW8012.ino

Add support for BL0937  to plugin _076_HLW8012
-Add possibility to set interrupt tigger for cf and cf1
Add possibility to set current reading state (low/high)

CaptainRoot committed:
add predefind pin settings for some energy meter sockets to HLW8012 / BL0937

CaptainRoot committed:
Revert changes to platformio.ini

CaptainRoot committed:
Update platformio.ini

TD-er committed:
Merge remote-tracking branch 'origin/feature/GPS_plugin' into feature/GPS_plugin

Grovkillen committed:
Merge pull request #3 from letscontrolit/mega

update from mega

Grovkillen committed:
[docs] P013 added

Grovkillen committed:
[docs] P003 added "EndOn"

Grovkillen committed:
[docs] P013 syntax fixed

TD-er committed:
[GPS] Make helper object to abstract hardware/software serial

TD-er committed:
[GPS] Extend helper object for abstracting HW/SW serial

TD-er committed:
[GPS] Serial wrapper support including ESP32

TD-er committed:
[GPS] typo in JSON file caused build issues

TD-er committed:
[core 2.3.0] build attempt (failing due to iRAM)

Attempt to test core stability using core 2.3.0.
It compiles, but linking is not possible, since iram1_0_seg is full.

TD-er committed:
[Serial] Let plugins use new serial wrapper

TD-er committed:
[WiFi] Set to auto reconnect.

TD-er committed:
[Serial] Use uniform serial port selector for all serial plugins

TD-er committed:
[Stability] Add delay(0) to all successful calls to plugin functions

Since these calls can be performed in sequences, the time of these calls can become rather large.
Every call returning success did something, so add a delay(0) after such a plugin function call to make sure background tasks are performed.

pecord committed:
pulled from the mq2 pr I found - https://github.com/letscontrolit/ESPEasy/pull/96
ended up pulling the values used for ppm from https://github.com/mdsiraj1992/Gassensors/blob/master/arduino%20codes/mq5_sensor/mq5_sensor.ino

pecord committed:
pulled from the mq2 pr I found - https://github.com/letscontrolit/ESPEasy/pull/96
ended up pulling the values used for ppm from https://github.com/mdsiraj1992/Gassensors/blob/master/arduino%20codes/mq5_sensor/mq5_sensor.ino

TD-er committed:
[build] Register ESPeasySerial as library in PlatformIO

TD-er committed:
[build] Added extra include of ESPeasySerial.h

Otherwise build depends on order of included files.

TD-er committed:
[#2197] Correct spelling error in 'Controller'

See #2197

TD-er committed:
[GPS] Allow GPS to update system time

Grovkillen committed:
[docs] P024 added

Grovkillen committed:
[docs] where used added for P024

Grovkillen committed:
[docs] P006 added

Grovkillen committed:
[docs] added note to BMP085 and BMP180

Grovkillen committed:
[docs] commands updated

Grovkillen committed:
[docs] P028 added

Grovkillen committed:
[docs] added info about bme/bmp similarities

Grovkillen committed:
[travis] -lstdc++-nox removed from esp82xx_2_5_0

Grovkillen committed:
[docs] fixed rule syntax typo

Grovkillen committed:
[flasher] getting ready for info splash with announcements

We could then use this to get in hold of more people than current sticky in the forum...

Grovkillen commented:
Change plugin number to 83 and name to `Gases - MQ5`. You can only have 4 values per plugin. So please add a drop-down to select which 4 wanted.

Grovkillen commented:
PS. I wonder when the plugin was lost in favor of the OLED (framed) plugin...

Grovkillen commented:
The status must be *TESTING* even though it was merged a long time ago (MQ2). And you need to add credit to the original developer!

Grovkillen commented:
And please have a look on the docs, you should add a manual section to the hardware. I will do the cleanup but you should address your thoughts so the end user understand you. Please add pictures of the hardware too. See template files here: https://www.letscontrolit.com/wiki/index.php?title=Template_Files

@TD-er we should move those templates to GitHub right? I got the source file (LaTeX) to it as well.

Grovkillen commented:
Travis is complaining about this: `PLATFORMIO_BUILD_FLAGS` error.

See my commit for fix: https://github.com/letscontrolit/ESPEasy/commit/f824c1a050b3e31219ace44a7e1c7ceee4b1085e

TD-er committed:
Merge pull request #2170 from giig1967g/bugfix/gpioEvents3

Bugfixes to GPIO handling and MONITOR command

TD-er committed:
Merge pull request #2191 from Grovkillen/UpdateDocs

[docs] P006, P013, P024, P028, P030 added.

psy0rz committed:
automatically updated release notes for mega-20181231

CaptainRoot committed:
movepredefind pin settings to PROGMEM

move predefind pin settings for some energy meter sockets to HLW8012 /BL0937 to PROGMEM

CaptainRoot committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

Grovkillen commented:
Will you do anything with this one or you expect us to do it?

TD-er committed:
[ESPEasy Serial] Fix build issues for core 2.4.0

TD-er committed:
[Puya] Backport core 2.5.0 Puya patch and make patched platform archives

This makes it possible to build Puya patched versions without having to patch the files first.

TD-er committed:
[JavaScript] Simple JS to update view in web interface on change serial

TD-er committed:
[GPS] Add PPS pin as reference pilot for time sync

TD-er committed:
[GPS] Add documentation for P082_GPS.ino

TD-er committed:
[PUYA] Use newest core 2.5.0 backport patch on unpatched sources.

TD-er committed:
[GPS] Include the GPS docs in the Spinx tree.

TD-er committed:
Merge pull request #2194 from TD-er/feature/GPS_plugin

P082: Feature/gps plugin & Serial port abstraction

TD-er committed:
[WiFi] Some basic fixes to improve wifi reconnect

As suggested by @clumsy-stefan  https://github.com/letscontrolit/ESPEasy/issues/1987#issuecomment-451644424

TD-er committed:
Merge pull request #2213 from TD-er/bugfix/1987_wifi_stability

[WiFi] Some basic fixes to improve wifi reconnect

psy0rz committed:
automatically updated release notes for mega-20190106

TD-er committed:
Add new Nodetype RPIEasy #2206

Also moved some String lookup functions and got rid of code duplication.

TD-er committed:
Merge pull request #2183 from CaptainRoot/mega

Update _P076_HLW8012.ino / Add support for BL0937 to plugin _076_HLW8012

TD-er committed:
Merge pull request #2216 from TD-er/feature/RPiEasy

Add new Nodetype RPIEasy #2206

TD-er committed:
[HLW8012] Fix some issues when selecting device

TD-er committed:
[HLW8012] Update documentation

TD-er committed:
Merge pull request #2218 from TD-er/bugfix/HLW8012

[HLW8012] Fix some issues when selecting device

TD-er commented:
Also good to have a link to [this site](http://wiki.seeedstudio.com/Grove-Gas_Sensor-MQ5/) in the documentation

TD-er committed:
[MQ5] Renamed plugin to P083

TD-er commented:
This plugin should be redesigned a lot.
It is taking several samples and then calls `delay` to wait for the next sample.
So calibration takes 50 samples and waits 500 msec per sample = 25 seconds completely blocking the system.
Taking a normal sample (5 samples, 50 msec interval) takes 250 msec in which the sensor will be blocking the whole system.

TD-er committed:
[MQ5] Changed plugin to use scheduler instead of delay

Also fixed some minor issues:
- Only 4 output values possible
- `break;` statement missing

TD-er committed:
Merge remote-tracking branch 'pecord/add-mq5-gas-sensor' into add-mq5-gas-sensor

TD-er commented:
Hmm, I think my rebase on top of the existing mega branch is not making this PR more clear...
I renamed the plugin, fixed some issues and used the scheduler instead of delays.

TD-er commented:
Close this one, but opened a new one: #2219
